### ISSUE TITLE
[Feature Request] - Add support for Lambda functions

### ISSUE THREAD

**Issue by aaron-russell** [alphagov/paas-roadmap#83]
**Is your feature request related to a problem? Please describe.**
We want to use Lambda to easily scale intensive parts of our application however currently we are not able to easily communicate with the database. as they are not part of the same space.

**Describe the outcome you'd like**
Support for a way to easily scale parts of the application either through lambda functions.

**Describe the solutions you've considered**
- cf tasks

**Comment by tlwr** [comment]
Have you looked at the recently released autoscaling feature, which is documented here: https://docs.cloud.service.gov.uk/managing_apps.html#autoscaling

**Comment by aaron-russell** [comment]
Yeah, that's great and we are looking at that separately, our service often has large files uploaded that are then processed. This may happen with serveral large users at once, but by the time we scaled, the instances would no longer be needed. So would be better if we could scale the upload part seperately and run a lmbda function for each request.

**Comment by tlwr** [comment]
Depending on how large "large" uploads are, you'll run into execution limits and gateway timeouts any sort of job. For example AWS Lambda was 5 minutes and is now 15 minutes.

Likewise CloudFront times out after 60 seconds, and our load balancers enforce a 900 second maximum request time (ALBs generally have 1hr-ish request timeout)

You might want to consider a queue instead, for which we have a separate roadmap item: https://github.com/alphagov/paas-roadmap/issues/16

**Comment by aaron-russell** [comment]
@tlwr we need to run the uploads instantly. We currently are using 10 instances but we are hitting limits with database connections so the code needs optimising more. We thought Lambdas could be a way of reducing the amount of instances as most of the time they are idle.

**Comment by tlwr** [comment]
In my opinion, functions aren't going to solve your database connection limits, as each function invocation will need a database connection, which will be less efficient than connection pooling done by a longer lived web application.

I would recommend using a web app to put the uploads onto a queue, and then having workers process from the queue.

**Comment by 46bit** [comment]
How heavily used are these database connections? You could increase the effective connection limit using pgbouncer—there’s a CF example of using it at https://github.com/cloudfoundry-samples/cf-ex-pgbouncer.

**Comment by aaron-russell** [comment]
@46bit Hello, our ORM already pools connections to the database. We have had to increase our max connections per instance up to 300 however we still have issues when updating large amounts of data (100,000+ rows)

**Comment by aaron-russell** [comment]
@tlwr Not at all, unfortunately I don't think our application would benefit from queuing as multipleuploads could be happening at the same time and the user will expect real time feedback on their files. Queuing these requests will make it slower for the user.

**Comment by tlwr** [comment]
Hi, I think I am not understanding your situation, perhaps because I have a different understanding of the terms we are using (eg queue) and I don't have a solid mental model of your application/service.

If your app is open source, would you be able to link to it, and the particular codepath that is causing your users issues (ie which controller/view/job). Some example timings would also be helpful (eg a file of 100k rows takes 4 minutes, etc)

**Comment by 46bit** [comment]
Hi @aaron-russell. We've talked about this within the GOV.UK PaaS team. So far there's two things we've identified from this:

1. It would help you for us to provide lambda functions or instant autoscaling. I'm afraid we aren't in a position to offer those in the foreseeable future. It's not something we can even have on our roadmap because there's so much uncertainty around it.
2. It would help you if our Postgres databases could be optionally accessible from outside of GOV.UK PaaS. That would let you run your own Lambdas with direct database access. We are looking at a Postgres supplier that would be accessible outside PaaS, but have no ETA on that.

For now, the best way to run Lambdas with database access is probably to run the Postgres yourself in RDS. We're happy to discuss that more at gov-uk-paas-support@digital.cabinet-office.gov.uk/
